library(GenomicInteractions)
library(rtracklayer)
library(GenomicFeatures)
library(ggbio)

#import topolically-associating domains (TAD) derived from 50-kilobase contact frequency matrix and idenified from chromatin interactions generated in human neural precursor cells (NPC)
hNSC_50kb_tad <- import("/Users/evangeller/Desktop/enhancer_analysis/hic_analysis/NPC_50kb_TAD.bed")

#import biological effects generated by genetic disruption of genes and conserved regions within cortical enhancers active in human neural stem cells
S0X_exp <- makeGRangesFromDataFrame(read.table("/Users/evangeller/Desktop/enhancer_analysis/mageck/S0X_mageck_annotate_classified_PCA.txt", header= TRUE), keep.extra.columns= TRUE)
S0X_phenotype <- S0X_exp[(S0X_exp$t12.phenotype == "negative") | (S0X_exp$t12.phenotype == "positive"),]

#import gencode annotation of genes
gencode <- import.gff("/Users/evangeller/Desktop/enhancer_analysis/bed/gencode.v19.annotation.gtf")
gencode_gene <- gencode[gencode$type == "gene"]
gencode_gene$id <- gencode_gene$gene_name
gencode_gene_phenotype <- gencode_gene[gencode_gene$gene_name %in% S0X_phenotype$name,]


expandRange = function(x, upstream=10000, downstream=10000) {
  strand_is_minus = strand(x) == "-"
  on_plus = which(!strand_is_minus)
  on_minus = which(strand_is_minus)
  start(x)[on_plus] = start(x)[on_plus] - upstream
  start(x)[on_minus] = start(x)[on_minus] - downstream
  end(x)[on_plus] = end(x)[on_plus] + downstream
  end(x)[on_minus] = end(x)[on_minus] + upstream
  x
}
#extend gencode annotations
gencode_gene_phenotype_promoter_gene <- expandRange(gencode_gene_phenotype)

#import enhancer-level summary of proliferation-altering regions
enhancer_set <- makeGRangesFromDataFrame(read.table( "/Users/evangeller/Desktop/enhancer_analysis/mageck/S0X_enhancer_annotate_obs.txt", header= TRUE), keep.extra.columns= TRUE)
enhancer_set$id <- as.character(granges(enhancer_set))
ac_set <- enhancer_set[enhancer_set$class == 'ac',]
ac_set_phenotype <- ac_set[ac_set$n.phenotypes.obs > 0,]
#extend H3K27ac enhancers 
start(ac_set_phenotype) <- start(ac_set_phenotype) - 10000
end(ac_set_phenotype) <- end(ac_set_phenotype) + 10000

#select enhancers that overlap with a TAD identified in human neural precursor cells
S0X_enhancer_assign <- ac_set_phenotype[ac_set_phenotype %over% hNSC_50kb_tad,]

#associate proliferation-altering enhancers with proliferation-altering genes based on co-localization within TAD boundaries
results <- GRanges()
for(i in seq_along(S0X_enhancer_assign)){
	local_domain <- hNSC_50kb_tad[hNSC_50kb_tad %over% S0X_enhancer_assign[i,]]
	local_gene_phenotypes <- gencode_gene_phenotype_promoter_gene[gencode_gene_phenotype_promoter_gene %over% local_domain]
	enhancer_gene_pair <- rep(S0X_enhancer_assign[i,], length(local_gene_phenotypes))
	enhancer_gene_pair$gene_name <- local_gene_phenotypes$id
	results <- c(results, enhancer_gene_pair)
}

#function association gene proliferation score with gene identified in enhancer-promoter chromatin interactions within TAD
enhancer_gene_PC1 <- function(enhancer) {
	gene_id <- enhancer$gene_name
	gene_phenotype <- S0X_phenotype[S0X_phenotype$name == gene_id,]
	return(gene_phenotype$PC1)
}

#apply enhancer_gene_PC1 function to granges object of enhancer-promoter chromatin interactions within TAD
mcols(results)$gene.PC1 <- t(sapply(results, enhancer_gene_PC1))

output <- as.data.frame(results, row.names = NULL)

write.table(output, "/Users/evangeller/Desktop/enhancer_analysis/hic_analysis/NPC_TAD_enhancer_gene_interaction.txt", row.names=F, sep=" ", quote= F)



